File: scripts/web_tool/browser_manager/browser_daemon.py
import asyncio
import json
import os
import socket
import sys
import threading
import time
from http.server import BaseHTTPRequestHandler, HTTPServer
from urllib.parse import urlparse, parse_qs

try:
    import websockets
except ImportError:
    print(json.dumps({"error": "websockets not installed"}))
    sys.exit(1)


def _find_free_port() -> int:
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind(("", 0))
        return s.getsockname()[1]


def _is_chrome_alive(pid: int) -> bool:
    try:
        import psutil
        return psutil.Process(pid).is_running()
    except Exception:
        return False


def _is_resource_type(resource_type: str) -> bool:
    return resource_type in ("Image", "Font", "Media", "Stylesheet", "Script",
                             "image", "font", "media", "stylesheet", "script")


def _now() -> str:
    return time.strftime("%Y-%m-%d %H:%M:%S") + f".{int(time.time() * 1000) % 1000:03d}"


def _get_browser_ws_url(cdp_port: int) -> str:
    import http.client
    conn = http.client.HTTPConnection("127.0.0.1", cdp_port, timeout=3)
    conn.request("GET", "/json/version")
    resp = conn.getresponse()
    data = json.loads(resp.read())
    conn.close()
    return data["webSocketDebuggerUrl"]


def _run_http_server(api_logs: list, resource_logs: list, console_logs: list,
                     capture_config: dict, save_dir: str, port: int, stop_event: threading.Event):
    class Handler(BaseHTTPRequestHandler):
        def _is_local(self):
            client_ip = self.client_address[0]
            return client_ip in ("127.0.0.1", "::1", "localhost")

        def _json_response(self, data, status=200):
            self.send_response(status)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps(data, ensure_ascii=False).encode())

        def do_GET(self):
            if not self._is_local():
                self.send_response(403)
                self.end_headers()
                return

            parsed = urlparse(self.path)
            path = parsed.path
            query = parse_qs(parsed.query)

            if path == "/api":
                limit = int(query.get("limit", ["10000"])[0])
                self._json_response(api_logs[-limit:])
            elif path == "/resources":
                limit = int(query.get("limit", ["10000"])[0])
                self._json_response(resource_logs[-limit:])
            elif path == "/console_log":
                limit = int(query.get("limit", ["10000"])[0])
                self._json_response(console_logs[-limit:])
            elif path == "/download-on":
                capture_config["download_resources"] = True
                self._json_response({"download_resources": True})
            elif path == "/download-off":
                capture_config["download_resources"] = False
                self._json_response({"download_resources": False})
            elif path == "/":
                self._json_response({
                    "endpoints": {
                        "/api": f"API 请求日志（最近 {len(api_logs)} 条）",
                        "/resources": f"资源请求元数据（最近 {len(resource_logs)} 条）",
                        "/console_log": f"Console 日志（最近 {len(console_logs)} 条）",
                        "/download-on": "开启资源下载（GET）",
                        "/download-off": "关闭资源下载（GET）",
                    },
                })
            else:
                self.send_response(404)
                self.end_headers()

        def do_POST(self):
            self.send_response(404)
            self.end_headers()

        def log_message(self, format, *args):
            pass

    server = HTTPServer(("", port), Handler)
    server.timeout = 1
    while not stop_event.is_set():
        server.handle_request()
    server.server_close()


async def _run_cdp_monitor(cdp_port: int, chrome_pid: int,
                           api_logs: list, resource_logs: list, console_logs: list,
                           capture_config: dict, save_dir: str, stop_event: threading.Event):
    browser_ws_url = _get_browser_ws_url(cdp_port)
    msg_id = 0
    frame_urls = {}
    pending_enables = []
    pending_resumes = []

    def _next_id():
        nonlocal msg_id
        msg_id += 1
        return msg_id

    async with websockets.connect(browser_ws_url, max_size=50 * 1024 * 1024) as ws:

        async def _send_cdp(method, params=None, session_id=None):
            msg = {"id": _next_id(), "method": method}
            if params:
                msg["params"] = params
            if session_id:
                msg["sessionId"] = session_id
            await ws.send(json.dumps(msg))

        # 1) setAutoAttach — 浏览器会在收到后发 Target.attachedToTarget 事件
        await _send_cdp("Target.setAutoAttach", {
            "autoAttach": True,
            "waitForDebuggerOnStart": True,
            "flatten": True,
        })

        # 2) 单一消息循环，处理所有 CDP 消息
        while not stop_event.is_set() and _is_chrome_alive(chrome_pid):
            try:
                raw = await asyncio.wait_for(ws.recv(), timeout=1)
            except asyncio.TimeoutError:
                # 处理 pending 的 enable/resume 任务
                while pending_enables:
                    sid = pending_enables.pop(0)
                    for domain in ["Network", "Runtime", "Log", "Page"]:
                        await _send_cdp(f"{domain}.enable", session_id=sid)
                while pending_resumes:
                    sid = pending_resumes.pop(0)
                    await _send_cdp("Runtime.runIfWaitingForDebugger", session_id=sid)
                continue
            except Exception:
                break

            try:
                data = json.loads(raw)
            except Exception:
                continue

            method = data.get("method", "")
            params = data.get("params", {})
            session_id = data.get("sessionId")

            if method == "Target.attachedToTarget":
                new_sid = params.get("sessionId", "")
                target_type = params.get("targetInfo", {}).get("type", "")
                if new_sid and target_type == "page":
                    pending_enables.append(new_sid)
                    pending_resumes.append(new_sid)
                    # 立即处理 enable
                    for domain in ["Network", "Runtime", "Log", "Page"]:
                        await _send_cdp(f"{domain}.enable", session_id=new_sid)
                    await _send_cdp("Runtime.runIfWaitingForDebugger", session_id=new_sid)
            else:
                _process_event(method, params, api_logs, resource_logs,
                               console_logs, capture_config, frame_urls)


def _process_event(method: str, params: dict,
                   api_logs: list, resource_logs: list, console_logs: list,
                   capture_config: dict, frame_urls: dict):
    if not method:
        return

    # ── Network.requestWillBeSent ──
    if method == "Network.requestWillBeSent":
        req = params.get("request", {})
        rt = params.get("type", "")
        frame_id = params.get("frameId", "")
        page_url = frame_urls.get(frame_id, "")

        if params.get("redirectResponse"):
            redir = params["redirectResponse"]
            entry = {
                "time": _now(),
                "page_url": page_url,
                "url": params.get("url", ""),
                "status": redir.get("status"),
                "resource_type": rt,
                "response_headers": redir.get("headers", {}),
            }
            if _is_resource_type(rt):
                resource_logs.append(entry)
                _trim(resource_logs, capture_config.get("resource_log_limit", 10000))
            else:
                api_logs.append(entry)
                _trim(api_logs, capture_config.get("api_log_limit", 10000))

        entry = {
            "time": _now(),
            "page_url": page_url,
            "url": req.get("url", ""),
            "method": req.get("method", ""),
            "resource_type": rt,
            "headers": req.get("headers", {}),
        }
        if req.get("postData"):
            entry["request_body"] = req["postData"]

        if _is_resource_type(rt):
            resource_logs.append(entry)
            _trim(resource_logs, capture_config.get("resource_log_limit", 10000))
        else:
            api_logs.append(entry)
            _trim(api_logs, capture_config.get("api_log_limit", 10000))

    # ── Network.responseReceived ──
    elif method == "Network.responseReceived":
        resp = params.get("response", {})
        rt = params.get("type", "")
        frame_id = params.get("frameId", "")
        page_url = frame_urls.get(frame_id, "")

        entry = {
            "time": _now(),
            "page_url": page_url,
            "url": resp.get("url", ""),
            "status": resp.get("status"),
            "resource_type": rt,
            "content_type": resp.get("headers", {}).get("content-type", ""),
            "content_length": resp.get("headers", {}).get("content-length"),
            "response_headers": resp.get("headers", {}),
        }

        if _is_resource_type(rt):
            resource_logs.append(entry)
            _trim(resource_logs, capture_config.get("resource_log_limit", 10000))
        else:
            api_logs.append(entry)
            _trim(api_logs, capture_config.get("api_log_limit", 10000))

    # ── Runtime.consoleAPICalled ──
    elif method == "Runtime.consoleAPICalled":
        args_list = params.get("args", [])
        text_parts = []
        for arg in args_list:
            if "value" in arg:
                text_parts.append(str(arg["value"]))
            elif "description" in arg:
                text_parts.append(arg["description"])
        console_logs.append({
            "time": _now(),
            "level": params.get("type", "log"),
            "text": " ".join(text_parts),
        })
        _trim(console_logs, capture_config.get("console_log_limit", 10000))

    # ── Runtime.exceptionThrown ──
    elif method == "Runtime.exceptionThrown":
        exc = params.get("exceptionDetails", {})
        text = exc.get("text", "")
        desc = exc.get("exception", {}).get("description", text)
        console_logs.append({
            "time": _now(),
            "level": "pageerror",
            "text": desc,
        })
        _trim(console_logs, capture_config.get("console_log_limit", 10000))

    # ── Page.frameNavigated ──
    elif method == "Page.frameNavigated":
        frame = params.get("frame", {})
        fid = frame.get("id", "")
        url = frame.get("url", "")
        if fid and url:
            frame_urls[fid] = url


def _trim(log_list: list, limit: int):
    while len(log_list) > limit:
        log_list.pop(0)


def main():
    args = sys.argv[1:]
    if len(args) < 3:
        print("用法: browser_daemon.py <cdp_port> <chrome_pid> <project_name>", file=sys.stderr)
        sys.exit(1)

    cdp_port = int(args[0])
    chrome_pid = int(args[1])
    project_name = args[2]

    api_logs = []
    resource_logs = []
    console_logs = []

    capture_config = {
        "download_resources": False,
        "body_limit": 1024 * 1024,
        "api_log_limit": 10000,
        "resource_log_limit": 10000,
        "console_log_limit": 10000,
    }

    stop_event = threading.Event()

    save_dir = os.path.join(os.path.dirname(__file__), "local_data", project_name)
    os.makedirs(save_dir, exist_ok=True)

    http_port = _find_free_port()
    http_thread = threading.Thread(
        target=_run_http_server,
        args=(api_logs, resource_logs, console_logs, capture_config, save_dir, http_port, stop_event),
        daemon=True,
    )
    http_thread.start()

    print(json.dumps({
        "api_url": f"http://127.0.0.1:{http_port}/api",
        "resources_url": f"http://127.0.0.1:{http_port}/resources",
        "console_log_url": f"http://127.0.0.1:{http_port}/console_log",
        "download_on_url": f"http://127.0.0.1:{http_port}/download-on",
        "download_off_url": f"http://127.0.0.1:{http_port}/download-off",
    }))
    sys.stdout.flush()

    asyncio.run(_run_cdp_monitor(cdp_port, chrome_pid, api_logs, resource_logs, console_logs,
                                 capture_config, save_dir, stop_event))


if __name__ == "__main__":
    main()

File: scripts/web_tool/browser_manager/browser_daemon.ts
import * as http from "node:http";
import * as net from "node:net";
import WebSocket from "ws";

const IS_RESOURCE_TYPE = new Set([
  "image", "font", "media", "stylesheet", "script",
  "Image", "Font", "Media", "Stylesheet", "Script",
]);

function now(): string {
  const d = new Date();
  const ms = d.getMilliseconds().toString().padStart(3, "0");
  const iso = d.toISOString().replace("T", " ").replace(/\.\d+Z$/, "");
  return `${iso}.${ms}`;
}

function trimList(list: unknown[], limit: number) {
  while (list.length > limit) list.shift();
}

function isAlive(pid: number): boolean {
  try { process.kill(pid, 0); return true; } catch { return false; }
}

async function main() {
  const args = process.argv.slice(2);
  if (args.length < 3) {
    process.stderr.write("用法: browser_daemon.ts <cdp_port> <chrome_pid> <project_name>\n");
    process.exit(1);
  }

  const cdpPort = parseInt(args[0]);
  const chromePid = parseInt(args[1]);
  const projectName = args[2];

  const apiLogs: Record<string, unknown>[] = [];
  const resourceLogs: Record<string, unknown>[] = [];
  const consoleLogs: Record<string, unknown>[] = [];

  const config = {
    download_resources: false,
    body_limit: 1024 * 1024,
    api_log_limit: 10000,
    resource_log_limit: 10000,
    console_log_limit: 10000,
  };

  // ── Find free port ──
  const httpPort = await new Promise<number>((resolve, reject) => {
    const s = net.createServer();
    s.listen(0, "127.0.0.1", () => {
      const port = (s.address() as net.AddressInfo).port;
      s.close(() => resolve(port));
    });
    s.on("error", reject);
  });

  // ── CDP Monitor via browser-level WebSocket ──
  const resp = await fetch(`http://127.0.0.1:${cdpPort}/json/version`);
  const versionData = await resp.json() as { webSocketDebuggerUrl: string };
  const browserWsUrl = versionData.webSocketDebuggerUrl;

  const ws = new WebSocket(browserWsUrl, { maxPayload: 50 * 1024 * 1024 });
  let msgId = 0;
  const frameUrls: Record<string, string> = {};
  const pendingCallbacks = new Map<number, { resolve: (data: Record<string, unknown>) => void; reject: (err: Error) => void }>();

  function nextId(): number { return ++msgId; }

  function sendCDP(method: string, params?: Record<string, unknown>, sessionId?: string) {
    const id = nextId();
    const msg: Record<string, unknown> = { id, method };
    if (params) msg.params = params;
    if (sessionId) msg.sessionId = sessionId;
    ws.send(JSON.stringify(msg));
    return id;
  }

  function sendCDPWithResponse(method: string, params?: Record<string, unknown>, sessionId?: string): Promise<Record<string, unknown>> {
    return new Promise((resolve, reject) => {
      const id = sendCDP(method, params, sessionId);
      pendingCallbacks.set(id, { resolve, reject });
      setTimeout(() => {
        if (pendingCallbacks.has(id)) {
          pendingCallbacks.delete(id);
          reject(new Error(`CDP timeout: ${method}`));
        }
      }, 10000);
    });
  }

  // pending body fetches: requestId -> { sessionId, entry }
  const pendingBodies = new Map<string, { sessionId?: string; entry: Record<string, unknown>; isRes: boolean }>();

  function processEvent(method: string, params: Record<string, unknown>, sessionId?: string) {
    if (!method) return;

    if (method === "Network.requestWillBeSent") {
      const req = (params.request || {}) as Record<string, unknown>;
      const rt = (params.type as string) || "";
      const frameId = (params.frameId as string) || "";
      const pageUrl = frameUrls[frameId] || "";
      const requestId = params.requestId as string;

      if (params.redirectResponse) {
        const redir = params.redirectResponse as Record<string, unknown>;
        const entry: Record<string, unknown> = {
          time: now(),
          page_url: pageUrl,
          url: params.url || "",
          status: redir.status,
          resource_type: rt,
          response_headers: redir.headers,
        };
        const isRes = IS_RESOURCE_TYPE.has(rt);
        const list = isRes ? resourceLogs : apiLogs;
        const limit = isRes ? config.resource_log_limit : config.api_log_limit;
        list.push(entry);
        trimList(list, limit);
      }

      const entry: Record<string, unknown> = {
        time: now(),
        page_url: pageUrl,
        url: req.url,
        method: req.method,
        resource_type: rt,
        headers: req.headers,
      };
      if (req.postData) entry.request_body = req.postData;

      const isRes = IS_RESOURCE_TYPE.has(rt);
      const list = isRes ? resourceLogs : apiLogs;
      const limit = isRes ? config.resource_log_limit : config.api_log_limit;
      list.push(entry);
      trimList(list, limit);

      // Track for response body fetching
      pendingBodies.set(requestId, { sessionId, entry, isRes });
    }

    else if (method === "Network.loadingFinished") {
      const requestId = params.requestId as string;
      const pending = pendingBodies.get(requestId);
      if (!pending) return;
      pendingBodies.delete(requestId);

      const { sessionId: sid, entry, isRes } = pending;
      // API requests: always fetch body; Resources: only if download enabled
      const shouldFetchBody = !isRes || config.download_resources;
      if (!shouldFetchBody) return;

      // Fire-and-forget body fetch — result handled in main message loop
      const bodyId = sendCDP("Network.getResponseBody", { requestId }, sid);
      pendingCallbacks.set(bodyId, {
        resolve: (result) => {
          const body = result.body as string;
          const base64 = result.base64Encoded as boolean;
          if (base64) {
            entry.response_body_base64 = body;
          } else if (body && body.length <= config.body_limit) {
            entry.response_body = body;
          } else if (body) {
            entry.response_body = body.slice(0, config.body_limit);
            entry.response_body_truncated = true;
          }
        },
        reject: () => { /* body fetch failed, ignore */ },
      });
      // Extend timeout for body fetches
      setTimeout(() => {
        if (pendingCallbacks.has(bodyId)) pendingCallbacks.delete(bodyId);
      }, 5000);
    }

    else if (method === "Network.responseReceived") {
      const respData = (params.response || {}) as Record<string, unknown>;
      const rt = (params.type as string) || "";
      const frameId = (params.frameId as string) || "";
      const pageUrl = frameUrls[frameId] || "";
      const headers = (respData.headers || {}) as Record<string, string>;
      const requestId = params.requestId as string;

      const entry: Record<string, unknown> = {
        time: now(),
        page_url: pageUrl,
        url: respData.url,
        status: respData.status,
        resource_type: rt,
        content_type: headers["content-type"] || "",
        content_length: headers["content-length"],
        response_headers: headers,
      };

      const isRes = IS_RESOURCE_TYPE.has(rt);
      const list = isRes ? resourceLogs : apiLogs;
      const limit = isRes ? config.resource_log_limit : config.api_log_limit;
      list.push(entry);
      trimList(list, limit);

      // Update pending body tracking with the response entry (prefer this one for body)
      if (pendingBodies.has(requestId)) {
        const pending = pendingBodies.get(requestId)!;
        pending.entry = entry;
      }
    }

    else if (method === "Runtime.consoleAPICalled") {
      const argsList = (params.args || []) as Array<Record<string, unknown>>;
      const textParts = argsList.map((arg) =>
        arg.value !== undefined ? String(arg.value) : (arg.description as string || "")
      );
      consoleLogs.push({
        time: now(),
        level: params.type || "log",
        text: textParts.join(" "),
      });
      trimList(consoleLogs, config.console_log_limit);
    }

    else if (method === "Runtime.exceptionThrown") {
      const excDetails = (params.exceptionDetails || {}) as Record<string, unknown>;
      const excObj = (excDetails.exception || {}) as Record<string, unknown>;
      const text = (excDetails.text || "") as string;
      consoleLogs.push({
        time: now(),
        level: "pageerror",
        text: (excObj.description as string) || text,
      });
      trimList(consoleLogs, config.console_log_limit);
    }

    else if (method === "Page.frameNavigated") {
      const frame = (params.frame || {}) as Record<string, unknown>;
      const fid = frame.id as string;
      const furl = frame.url as string;
      if (fid && furl) frameUrls[fid] = furl;
    }
  }

  // Wait for WebSocket to open
  await new Promise<void>((resolve, reject) => {
    ws.on("open", resolve);
    ws.on("error", reject);
  });

  // setAutoAttach + setDiscoverTargets for existing/initial pages
  sendCDP("Target.setDiscoverTargets", { discover: true });
  sendCDP("Target.setAutoAttach", {
    autoAttach: true,
    waitForDebuggerOnStart: true,
    flatten: true,
  });

  // Track attached sessions
  const attachedSessions = new Map<string, string>(); // sessionId -> targetId

  // Enable domains on a session (fire-and-forget for setAutoAttach caught pages)
  function enableDomainsOnSession(sessionId: string) {
    for (const domain of ["Network", "Runtime", "Log", "Page"]) {
      sendCDP(`${domain}.enable`, undefined, sessionId);
    }
  }

  // Main message loop
  ws.on("message", (raw: Buffer) => {
    let data: Record<string, unknown>;
    try { data = JSON.parse(raw.toString()); } catch { return; }

    // Handle CDP command responses
    if (data.id && pendingCallbacks.has(data.id as number)) {
      const cb = pendingCallbacks.get(data.id as number)!;
      pendingCallbacks.delete(data.id as number);
      if (data.error) {
        cb.reject(new Error(JSON.stringify(data.error)));
      } else {
        cb.resolve((data.result || {}) as Record<string, unknown>);
      }
      return;
    }

    const method = data.method as string;
    const params = (data.params || {}) as Record<string, unknown>;

    if (method === "Target.attachedToTarget") {
      const newSid = params.sessionId as string;
      const targetInfo = (params.targetInfo || {}) as Record<string, unknown>;
      const targetType = targetInfo.type as string;
      const targetId = targetInfo.targetId as string;

      if (newSid && targetType === "page") {
        attachedSessions.set(newSid, targetId);
        enableDomainsOnSession(newSid);
        sendCDP("Runtime.runIfWaitingForDebugger", undefined, newSid);
        // Recursive setAutoAttach for sub-targets (iframes, workers)
        sendCDP("Target.setAutoAttach", {
          autoAttach: true,
          waitForDebuggerOnStart: true,
          flatten: true,
        }, newSid);
      } else if (newSid) {
        // Non-page targets (workers, etc.) — just resume
        sendCDP("Runtime.runIfWaitingForDebugger", undefined, newSid);
      }
    } else if (method) {
      const sessionId = data.sessionId as string | undefined;
      processEvent(method, params, sessionId);
    }
  });

  // ── Open new tab via CDP (createTarget → attach → enable → navigate) ──
  // This is the reliable path that guarantees Network.enable before navigation
  async function openNewTab(url: string): Promise<{ targetId: string }> {
    const result = await sendCDPWithResponse("Target.createTarget", { url: "about:blank" });
    const targetId = result.targetId as string;

    const attachResult = await sendCDPWithResponse("Target.attachToTarget", {
      targetId,
      flatten: true,
    });
    const sessionId = attachResult.sessionId as string;
    attachedSessions.set(sessionId, targetId);

    enableDomainsOnSession(sessionId);
    sendCDP("Runtime.runIfWaitingForDebugger", undefined, sessionId);

    // Navigate after domains are enabled
    await sendCDPWithResponse("Page.navigate", { url }, sessionId);

    return { targetId };
  }

  // Navigate in existing tab
  async function navigateTab(tabIndex: number, url: string): Promise<void> {
    const targetsResp = await fetch(`http://127.0.0.1:${cdpPort}/json/list`);
    const targets = (await targetsResp.json() as Array<Record<string, unknown>>)
      .filter((t) => t.type === "page");

    if (tabIndex < 0 || tabIndex >= targets.length) {
      throw new Error(`标签页索引 ${tabIndex} 超出范围，当前共 ${targets.length} 个标签页`);
    }

    const targetId = targets[tabIndex].id as string;
    // Find the session for this target
    let sessionId: string | undefined;
    for (const [sid, tid] of attachedSessions) {
      if (tid === targetId) { sessionId = sid; break; }
    }

    if (sessionId) {
      await sendCDPWithResponse("Page.navigate", { url }, sessionId);
    } else {
      // Not attached yet, attach first
      const attachResult = await sendCDPWithResponse("Target.attachToTarget", {
        targetId,
        flatten: true,
      });
      sessionId = attachResult.sessionId as string;
      attachedSessions.set(sessionId, targetId);
      enableDomainsOnSession(sessionId);
      sendCDP("Runtime.runIfWaitingForDebugger", undefined, sessionId);
      await sendCDPWithResponse("Page.navigate", { url }, sessionId);
    }
  }

  // ── HTTP Server ──
  const server = http.createServer((req, res) => {
    const clientIp = req.socket.remoteAddress || "";
    if (!["127.0.0.1", "::1", "::ffff:127.0.0.1"].includes(clientIp)) {
      res.writeHead(403, { "Content-Type": "application/json" });
      res.end('{"error":"forbidden"}');
      return;
    }

    const url = new URL(req.url || "/", `http://127.0.0.1:${httpPort}`);
    const path = url.pathname;
    const limit = parseInt(url.searchParams.get("limit") || "10000");

    const jsonRes = (data: unknown, status = 200) => {
      res.writeHead(status, { "Content-Type": "application/json" });
      res.end(JSON.stringify(data));
    };

    if (path === "/api") { jsonRes(apiLogs.slice(-limit)); }
    else if (path === "/resources") { jsonRes(resourceLogs.slice(-limit)); }
    else if (path === "/console_log") { jsonRes(consoleLogs.slice(-limit)); }
    else if (path === "/download-on") { config.download_resources = true; jsonRes({ download_resources: true }); }
    else if (path === "/download-off") { config.download_resources = false; jsonRes({ download_resources: false }); }
    else if (path === "/open-tab") {
      const targetUrl = url.searchParams.get("url") || "";
      const tabIdx = url.searchParams.has("tab_index") ? parseInt(url.searchParams.get("tab_index")!) : undefined;
      if (!targetUrl) { jsonRes({ error: "missing url param" }, 400); return; }

      const handler = tabIdx !== undefined
        ? navigateTab(tabIdx, targetUrl)
        : openNewTab(targetUrl);

      handler
        .then((result) => jsonRes({ ok: true, ...result }))
        .catch((err: Error) => jsonRes({ error: err.message }, 500));
    }
    else if (path === "/") {
      jsonRes({
        endpoints: {
          "/api": `API 请求日志（最近 ${apiLogs.length} 条）`,
          "/resources": `资源请求元数据（最近 ${resourceLogs.length} 条）`,
          "/console_log": `Console 日志（最近 ${consoleLogs.length} 条）`,
          "/open-tab": "在新标签页打开 URL（?url=...&tab_index=...）",
          "/download-on": "开启资源下载（GET）",
          "/download-off": "关闭资源下载（GET）",
        },
      });
    } else {
      res.writeHead(404);
      res.end();
    }
  });

  server.listen(httpPort, "127.0.0.1");

  // ── Output daemon info ──
  const daemonInfo = {
    api_url: `http://127.0.0.1:${httpPort}/api`,
    resources_url: `http://127.0.0.1:${httpPort}/resources`,
    console_log_url: `http://127.0.0.1:${httpPort}/console_log`,
    open_tab_url: `http://127.0.0.1:${httpPort}/open-tab`,
    download_on_url: `http://127.0.0.1:${httpPort}/download-on`,
    download_off_url: `http://127.0.0.1:${httpPort}/download-off`,
  };
  console.log(JSON.stringify(daemonInfo));

  // Keep alive until browser dies
  await new Promise<void>((resolve) => {
    const check = setInterval(() => {
      if (!isAlive(chromePid)) {
        clearInterval(check);
        ws.close();
        resolve();
      }
    }, 2000);
  });

  server.close();
  process.exit(0);
}

main().catch((err) => {
  process.stderr.write(`daemon error: ${err}\n`);
  process.exit(1);
});
